src/public/js/rs.js

```diff
@@ -45,13 +45,208 @@ var RS = (function() {
+      var repairedErrors = false;
+      for (var i = 0; i < n - k; i++) {
+        if (repairedRemainder[i] != 0) {
+          throw new Error('Too many errors to decode.');
+        }
+      }
+    }
+    return receivedMsg.slice(0, k);
+  }
+
+  // Computes a syndrome polynomial (the sum from i = 0 to n-k of Si*X^i)
+  // for the specified received message by evaluating the received message
+  // at n-k consecutive powers of the generator element within Galois Field 256
+  // receivedMsg is an array of length n
+  // n specifies the length of the codeword / received message
+  // k specifies the length of the original message
+  // Returns a big-endian array of length (n-k+1) which represents a polynomial
+  // -> [1, 2, 3] is x^2 + 2*x + 3
+  function computeSyndromePoly(receivedMsg, n, k) {
+    var syndromes = new Uint8Array(n - k + 1);
+
+    // Add the 0th syndrome to represent the coefficient of z^0 even though it's not useful for error correction
+    syndromes[n - k] = 0;
+
+    var syndrome;
+    for (var i = 1; i <= n - k; i++) {
+      syndrome = UTIL.polynomialEval(receivedMsg, UTIL.exp3(i));
+      syndromes[syndromes.length - 1 - i] = syndrome;
     }
+
+    return syndromes;
   }
 
+  // Performs the Berlekamp-Massey algorithm on a syndrome polynomial for
+  // a certain n and k. Iterates (n-k) times to fit the BM "key equation".
+  // Returns an error location polynomial and an error magnitude polynomial.
+  // Input syndromePoly and returns errorLocationPoly and errorEvaluatorPoly are
+  // big-endian arrays representing polynomials, i.e. [1,2,3] -> x^2 + 2x + 3
+  // Inputs n and k are integers
+  function berlekampMassey(syndromePoly, n, k) {
+    // Compute 1 + syndromePoly in GF (copying to avoid modifying input)
+    // Changes the 0 in the z^0 place in syndromePoly to a 1, which we need mathematically
+    var syndromePolyPlusOne = syndromePoly.slice(0);
+    syndromePolyPlusOne[syndromePoly.length - 1] ^= 1;
+
+    // Initial conditions
+    var sigmas = new Array(n - k + 1);
+    var omegas = new Array(n - k + 1);
+    var taus = new Array(n - k + 1);
+    var gammas = new Array(n - k + 1);
+    var Ds = new Array(n - k + 1);
+    var Bs = new Array(n - k + 1);
+    sigmas[0] = new Uint8Array([1]);
+    omegas[0] = new Uint8Array([1]);
+    taus[0] = new Uint8Array([1]);
+    gammas[0] = new Uint8Array([0]);
+    Ds[0] = 0;
+    Bs[0] = false;
+
+    // Perform n-k iterations
+    var lhs;
+    var delta;
+    var shiftedTau;
+    var shiftedGamma;
+    var scaledTau;
+    var scaledGamma;
+    var DThreshold;
+    for (var i = 0; i <= n - k; i++) {
+      // Compute the left-hand side of the key equation
+      lhs = UTIL.polynomialMult(sigmas[i], syndromePolyPlusOne);
+
+      // Delta is the coefficient of z^(i+1)
+      delta = lhs[(lhs.length - 1) - (i + 1)];
+
+      // Use tau and gamma to update sigma and omega:
+      shiftedTau = UTIL.mergeTypedArrays(taus[i], new Uint8Array([0])); // z * tau
```

Comment:
also trying to think -- this might be a place where pointer-map arrays would be faster (all we're doing is adding a zero at the end and pointer-map arrays have a .push method) but pretty sure the typed arrays are overall more performant